File: Sources/App/Classes/IRC/IRCDirectChatConnection.m
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\___/_/\_\\__|\__,_|\__,_|_|
 *
 * Copyright (c) 2008 - 2010 Satoshi Nakagawa <psychs AT limechat DOT net>
 * Copyright (c) 2010 - 2018 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of Textual, "Codeux Software, LLC", nor the
 *    names of its contributors may be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *********************************************************************** */

#import "IRCClientPrivate.h"
#import "IRCDirectChatConnectionPrivate.h"
#import "TDCFileTransferDialogSocketPrivate.h"
#import "TPCPreferencesLocal.h"

NS_ASSUME_NONNULL_BEGIN

#define _connectTimeout 30.0
#define _writeTimeout 30.0

/* A single line is never allowed to grow without bound. */
#define _maximumLineLength (1024 * 16)

#define _assertMainQueue() dispatch_assert_queue(dispatch_get_main_queue())

@interface IRCDirectChatConnection () <TDCFileTransferDialogSocketDelegate>
@property(nonatomic, weak, readwrite) IRCClient *client;
@property(nonatomic, weak, readwrite, nullable) id<IRCDirectChatConnectionDelegate> delegate;
@property(nonatomic, copy, readwrite) NSString *peerNickname;
@property(nonatomic, copy, readwrite, nullable) NSString *hostAddress;
@property(nonatomic, assign, readwrite) uint16_t hostPort;
@property(nonatomic, copy, readwrite, nullable) NSString *transferToken;
@property(nonatomic, assign, readwrite) IRCDirectChatConnectionState state;
@property(nonatomic, assign) BOOL isListener;
@property(nonatomic, strong, nullable) TDCFileTransferDialogSocket *listeningServer;
@property(nonatomic, strong, nullable) TDCFileTransferDialogSocket *connection;
@property(nonatomic, strong, nullable) XRPortMapper *portMapping;
@property(nonatomic, strong) NSMutableData *lineBuffer;
@end

@implementation IRCDirectChatConnection

#pragma mark -
#pragma mark Initialization

+ (instancetype)connectionToPeer:(NSString *)nickname
						 address:(NSString *)hostAddress
							port:(uint16_t)hostPort
						onClient:(IRCClient *)client
						delegate:(id<IRCDirectChatConnectionDelegate>)delegate
{
	NSParameterAssert(hostAddress != nil);
	NSParameterAssert(hostPort != 0);

	IRCDirectChatConnection *object = [[self alloc] initWithPeer:nickname onClient:client delegate:delegate];

	object.hostAddress = hostAddress;
	object.hostPort = hostPort;

	object.isListener = NO;

	return object;
}

+ (instancetype)listeningConnectionForPeer:(NSString *)nickname
									 token:(nullable NSString *)transferToken
								  onClient:(IRCClient *)client
								  delegate:(id<IRCDirectChatConnectionDelegate>)delegate
{
	IRCDirectChatConnection *object = [[self alloc] initWithPeer:nickname onClient:client delegate:delegate];

	if (transferToken.length > 0) {
		object.transferToken = transferToken;
	}

	object.isListener = YES;

	return object;
}

- (instancetype)initWithPeer:(NSString *)nickname
					onClient:(IRCClient *)client
					delegate:(id<IRCDirectChatConnectionDelegate>)delegate
{
	NSParameterAssert(nickname != nil);
	NSParameterAssert(client != nil);
	NSParameterAssert(delegate != nil);

	if ((self = [super init])) {
		self.peerNickname = nickname;
		self.client = client;
		self.delegate = delegate;

		self.lineBuffer = [NSMutableData data];

		self.state = IRCDirectChatConnectionStateIdle;

		return self;
	}

	return nil;
}

- (void)dealloc
{
	[self tearDownSockets];
}

#pragma mark -
#pragma mark Properties

- (BOOL)isConnected
{
	return (self.state == IRCDirectChatConnectionStateConnected);
}

#pragma mark -
#pragma mark Open/Close

- (void)open
{
	_assertMainQueue();

	if (self.state != IRCDirectChatConnectionStateIdle) {
		return;
	}

	if (self.isListener) {
		[self openListener];
	} else {
		[self openConnection];
	}
}

- (void)openConnection
{
	self.state = IRCDirectChatConnectionStateConnecting;

	TDCFileTransferDialogSocket *connection =
		[[TDCFileTransferDialogSocket alloc] initWithDelegate:self delegateQueue:dispatch_get_main_queue()];

	self.connection = connection;

	[connection connectToHost:self.hostAddress
						 port:self.hostPort
				 viaInterface:[TPCPreferences fileTransferIPAddressInterfaceName]
					  timeout:_connectTimeout];
}

- (void)openListener
{
	uint16_t portRangeStart = [TPCPreferences fileTransferPortRangeStart];
	uint16_t portRangeEnd = [TPCPreferences fileTransferPortRangeEnd];

	if (portRangeStart == 0 || portRangeStart > portRangeEnd) {
		[self closeWithError:[TDCFileTransferDialogSocket errorWithCode:TDCFileTransferDialogSocketErrorNoOpenPort
															description:@"The file transfer port range is invalid"]];

		return;
	}

	self.state = IRCDirectChatConnectionStateListening;

	TDCFileTransferDialogSocket *listeningServer =
		[[TDCFileTransferDialogSocket alloc] initWithDelegate:self delegateQueue:dispatch_get_main_queue()];

	self.listeningServer = listeningServer;

	[listeningServer listenOnPortRangeFrom:portRangeStart to:portRangeEnd];
}

- (void)close
{
	_assertMainQueue();

	if (self.state == IRCDirectChatConnectionStateClosed) {
		return;
	}

	self.state = IRCDirectChatConnectionStateClosed;

	[self tearDownSockets];
}

- (void)closeWithError:(nullable NSError *)error
{
	if (self.state == IRCDirectChatConnectionStateClosed) {
		return;
	}

	[self close];

	[self.delegate directChatConnection:self didCloseWithError:error];
}

- (void)tearDownSockets
{
	XRPortMapper *portMapping = self.portMapping;

	if (portMapping) {
		[RZNotificationCenter() removeObserver:self name:XRPortMapperDidChangedNotification object:portMapping];

		self.portMapping = nil;

		[portMapping close];
	}

	TDCFileTransferDialogSocket *listeningServer = self.listeningServer;

	if (listeningServer) {
		self.listeningServer = nil;

		[listeningServer disconnect];
	}

	TDCFileTransferDialogSocket *connection = self.connection;

	if (connection) {
		self.connection = nil;

		[connection disconnect];
	}
}

#pragma mark -
#pragma mark Port Mapping

/* Mirrors the file transfer behaviour: try to map the port through the
 router, but advertise the listener either way. */
- (void)mapListeningPort:(uint16_t)port
{
	XRPortMapper *portMapping = [[XRPortMapper alloc] initWithPort:port];

	portMapping.mapTCP = YES;
	portMapping.mapUDP = NO;

	portMapping.desiredPublicPort = port;

	self.portMapping = portMapping;

	[RZNotificationCenter() addObserver:self
							   selector:@selector(portMapperDidFinishWork:)
								   name:XRPortMapperDidChangedNotification
								 object:portMapping];

	if ([portMapping open] == NO) {
		[self portMapperDidFinishWork:nil];
	}
}

- (void)portMapperDidFinishWork:(nullable NSNotification *)aNotification
{
	_assertMainQueue();

	if (self.state != IRCDirectChatConnectionStateListening) {
		return;
	}

	XRPortMapper *portMapping = self.portMapping;

	if (portMapping == nil) {
		return;
	}

	[RZNotificationCenter() removeObserver:self name:XRPortMapperDidChangedNotification object:portMapping];

	if (portMapping.isMapped) {
		LogToConsoleInfo("Direct chat: port %{public}hu mapped", self.hostPort);
	} else {
		LogToConsoleError("Direct chat: port mapping failed with error code %{public}i", portMapping.error);
	}

	[self.delegate directChatConnection:self didStartListeningOnPort:self.hostPort];
}

#pragma mark -
#pragma mark Sending

- (void)sendMessage:(NSString *)message
{
	NSParameterAssert(message != nil);

	[self sendLine:message];
}

- (void)sendAction:(NSString *)message
{
	NSParameterAssert(message != nil);

	[self sendLine:[NSString stringWithFormat:@"%cACTION %@%c", 0x01, message, 0x01]];
}

- (void)sendLine:(NSString *)line
{
	_assertMainQueue();

	if (self.isConnected == NO) {
		return;
	}

	/* A newline inside the text would be read by the peer as two
	 messages. The caller already split on newlines; this is a guard. */
	NSString *sanitizedLine = [line stringByReplacingOccurrencesOfString:@"\r" withString:@" "];

	sanitizedLine = [sanitizedLine stringByReplacingOccurrencesOfString:@"\n" withString:@" "];

	NSMutableData *data = [[self.client convertToCommonEncoding:sanitizedLine] mutableCopy];

	if (data == nil) {
		return;
	}

	[data appendBytes:"\r\n" length:2];

	[self.connection writeData:data timeout:_writeTimeout];
}

#pragma mark -
#pragma mark Receiving

- (void)consumeReceivedData:(NSData *)data
{
	NSMutableData *lineBuffer = self.lineBuffer;

	[lineBuffer appendData:data];

	while (lineBuffer.length > 0) {
		NSRange newlineRange = [lineBuffer rangeOfData:[NSData dataWithBytes:"\n" length:1]
											   options:0
												 range:NSMakeRange(0, lineBuffer.length)];

		if (newlineRange.location == NSNotFound) {
			if (lineBuffer.length > _maximumLineLength) {
				[self closeWithError:[TDCFileTransferDialogSocket
										 errorWithCode:TDCFileTransferDialogSocketErrorBadParameter
										   description:@"The peer sent a line which is too long"]];
			}

			return;
		}

		NSData *lineData = [lineBuffer subdataWithRange:NSMakeRange(0, newlineRange.location)];

		[lineBuffer replaceBytesInRange:NSMakeRange(0, newlineRange.location + 1) withBytes:NULL length:0];

		[self consumeReceivedLine:lineData];

		/* The delegate may have closed the chat in response. */
		if (self.state == IRCDirectChatConnectionStateClosed) {
			return;
		}
	}
}

- (void)consumeReceivedLine:(NSData *)lineData
{
	NSString *line = [self.client convertFromCommonEncoding:lineData];

	if (line == nil) {
		return;
	}

	line = [line stringByTrimmingCharactersInSet:[NSCharacterSet newlineCharacterSet]];

	if (line.length == 0) {
		return;
	}

	BOOL isAction = NO;

	NSString *actionPrefix = [NSString stringWithFormat:@"%cACTION ", 0x01];

	if ([line hasPrefix:actionPrefix]) {
		NSUInteger prefixLength = actionPrefix.length;

		NSUInteger suffixLength = ([line hasSuffix:[NSString stringWithFormat:@"%c", 0x01]] ? 1 : 0);

		if (line.length > (prefixLength + suffixLength)) {
			line = [line substringWithRange:NSMakeRange(prefixLength, (line.length - prefixLength - suffixLength))];
		} else {
			line = @"";
		}

		isAction = YES;
	}

	[self.delegate directChatConnection:self didReceiveMessage:line isAction:isAction];
}

#pragma mark -
#pragma mark Socket Delegate

- (void)socket:(TDCFileTransferDialogSocket *)socket didStartListeningOnPort:(uint16_t)port
{
	_assertMainQueue();

	if (socket != self.listeningServer) {
		return;
	}

	self.hostPort = port;

	[self mapListeningPort:port];
}

- (void)socket:(TDCFileTransferDialogSocket *)socket didFailToListenWithError:(NSError *)error
{
	_assertMainQueue();

	if (socket != self.listeningServer) {
		return;
	}

	[self closeWithError:error];
}

- (void)socket:(TDCFileTransferDialogSocket *)socket didAcceptConnection:(TDCFileTransferDialogSocket *)connection
{
	_assertMainQueue();

	if (socket != self.listeningServer || self.state != IRCDirectChatConnectionStateListening) {
		[connection disconnect];

		return;
	}

	/* One peer only. Stop listening the moment somebody connects. */
	self.listeningServer = nil;

	[socket disconnect];

	self.connection = connection;

	[self connectionDidBecomeReady];
}

- (void)socketDidConnect:(TDCFileTransferDialogSocket *)socket
{
	_assertMainQueue();

	if (socket != self.connection || self.state != IRCDirectChatConnectionStateConnecting) {
		return;
	}

	[self connectionDidBecomeReady];
}

- (void)connectionDidBecomeReady
{
	self.state = IRCDirectChatConnectionStateConnected;

	[self.connection readData];

	[self.delegate directChatConnectionDidConnect:self];
}

- (void)socket:(TDCFileTransferDialogSocket *)socket didReadData:(NSData *)data
{
	_assertMainQueue();

	if (socket != self.connection || self.isConnected == NO) {
		return;
	}

	[self consumeReceivedData:data];

	if (self.isConnected) {
		[socket readData];
	}
}

- (void)socketDidWriteData:(TDCFileTransferDialogSocket *)socket
{
	/* Nothing to do. Writes are fire and forget. */
}

- (void)socket:(TDCFileTransferDialogSocket *)socket didDisconnectWithError:(nullable NSError *)error
{
	_assertMainQueue();

	if (socket != self.connection && socket != self.listeningServer) {
		return;
	}

	/* A clean close by the peer arrives as a "closed by peer" error.
	 Treat that as a normal end of the conversation. */
	if ([error.domain isEqualToString:TDCFileTransferDialogSocketErrorDomain] &&
		error.code == TDCFileTransferDialogSocketErrorClosedByPeer) {
		error = nil;
	}

	[self closeWithError:error];
}

@end

NS_ASSUME_NONNULL_END

File: Sources/App/Modules/IRCProtocol/DirectChat/IRCDirectChatConnection.swift
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \\ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\\___/_/\\_\\__|\\__,_|\\__,_|_|
 *
 * Copyright (c) 2008 - 2010 Satoshi Nakagawa <psychs AT limechat DOT net>
 * Copyright (c) 2010 - 2026 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 *********************************************************************** */

import Foundation
import os

private let connectTimeout: TimeInterval = 30.0
private let writeTimeout: TimeInterval = 30.0
private let maximumLineLength = 1024 * 16

private let directChatLogger = Logger(
	subsystem: Bundle.main.bundleIdentifier ?? "Glasstual",
	category: "DirectChat"
)

@objc
public enum IRCDirectChatConnectionState: UInt {
	case idle = 0
	case listening
	case connecting
	case connected
	case closed
}

@objc
public protocol IRCDirectChatConnectionDelegate: NSObjectProtocol {
	@objc(directChatConnection:didStartListeningOnPort:)
	func directChatConnection(_ connection: DirectChatConnection, didStartListeningOnPort port: UInt16)

	@objc(directChatConnectionDidConnect:)
	func directChatConnectionDidConnect(_ connection: DirectChatConnection)

	@objc(directChatConnection:didReceiveMessage:isAction:)
	func directChatConnection(
		_ connection: DirectChatConnection,
		didReceiveMessage message: String,
		isAction: Bool
	)

	@objc(directChatConnection:didCloseWithError:)
	func directChatConnection(_ connection: DirectChatConnection, didCloseWithError error: Error?)
}

/* A DCC CHAT session. One instance is one TCP connection to one peer.
 All state belongs to the main queue: the socket delivers its callbacks
 there and every method below must be called from there. */
@objc(IRCDirectChatConnection)
public final class DirectChatConnection: NSObject, TDCFileTransferDialogSocketDelegate {
	@objc public private(set) weak var client: IRCClient?
	@objc public private(set) weak var delegate: IRCDirectChatConnectionDelegate?
	@objc public private(set) var peerNickname: String
	@objc public private(set) var hostAddress: String?
	@objc public private(set) var hostPort: UInt16 = 0
	@objc public private(set) var transferToken: String?
	@objc public private(set) var state: IRCDirectChatConnectionState = .idle

	private var isListener = false
	private var listeningServer: TDCFileTransferDialogSocket?
	private var connection: TDCFileTransferDialogSocket?
	private var portMapping: XRPortMapper?
	private var lineBuffer = NSMutableData()

	@objc public var isConnected: Bool {
		state == .connected
	}

	@available(*, unavailable)
	override public init() {
		fatalError("Use factory methods")
	}

	private init(
		peer nickname: String,
		onClient client: IRCClient,
		delegate: IRCDirectChatConnectionDelegate
	) {
		peerNickname = nickname
		self.client = client
		self.delegate = delegate
		super.init()
	}

	deinit {
		tearDownSockets()
	}

	@objc(connectionToPeer:address:port:onClient:delegate:)
	public class func connection(
		toPeer nickname: String,
		address hostAddress: String,
		port hostPort: UInt16,
		onClient client: IRCClient,
		delegate: IRCDirectChatConnectionDelegate
	) -> DirectChatConnection {
		precondition(hostPort != 0)

		let object = DirectChatConnection(peer: nickname, onClient: client, delegate: delegate)
		object.hostAddress = hostAddress
		object.hostPort = hostPort
		object.isListener = false
		return object
	}

	@objc(listeningConnectionForPeer:token:onClient:delegate:)
	public class func listeningConnection(
		forPeer nickname: String,
		token transferToken: String?,
		onClient client: IRCClient,
		delegate: IRCDirectChatConnectionDelegate
	) -> DirectChatConnection {
		let object = DirectChatConnection(peer: nickname, onClient: client, delegate: delegate)

		if let transferToken, transferToken.isEmpty == false {
			object.transferToken = transferToken
		}

		object.isListener = true
		return object
	}

	@objc
	public func open() {
		dispatchPrecondition(condition: .onQueue(.main))

		guard state == .idle else {
			return
		}

		if isListener {
			openListener()
		} else {
			openConnection()
		}
	}

	private func openConnection() {
		state = .connecting

		let connection = TDCFileTransferDialogSocket(
			delegate: self,
			delegateQueue: DispatchQueue.main
		)
		self.connection = connection

		guard let hostAddress else {
			return
		}

		connection.connect(
			toHost: hostAddress,
			port: hostPort,
			viaInterface: TPCPreferences.fileTransferIPAddressInterfaceName(),
			timeout: connectTimeout
		)
	}

	private func openListener() {
		let portRangeStart = TPCPreferences.fileTransferPortRangeStart()
		let portRangeEnd = TPCPreferences.fileTransferPortRangeEnd()

		if portRangeStart == 0 || portRangeStart > portRangeEnd {
			close(
				with: TDCFileTransferDialogSocket.error(
					withCode: .noOpenPort,
					description: "The file transfer port range is invalid"
				)
			)
			return
		}

		state = .listening

		let listeningServer = TDCFileTransferDialogSocket(
			delegate: self,
			delegateQueue: DispatchQueue.main
		)
		self.listeningServer = listeningServer
		listeningServer.listenOnPortRange(from: portRangeStart, to: portRangeEnd)
	}

	@objc
	public func close() {
		dispatchPrecondition(condition: .onQueue(.main))

		guard state != .closed else {
			return
		}

		state = .closed
		tearDownSockets()
	}

	private func close(with error: Error?) {
		guard state != .closed else {
			return
		}

		close()
		delegate?.directChatConnection(self, didCloseWithError: error)
	}

	private func tearDownSockets() {
		if let portMapping {
			NotificationCenter.default.removeObserver(
				self,
				name: .XRPortMapperDidChanged,
				object: portMapping
			)
			self.portMapping = nil
			portMapping.close()
		}

		if let listeningServer {
			self.listeningServer = nil
			listeningServer.disconnect()
		}

		if let connection {
			self.connection = nil
			connection.disconnect()
		}
	}

	/* Mirrors the file transfer behaviour: try to map the port through the
	 router, but advertise the listener either way. */
	private func mapListeningPort(_ port: UInt16) {
		let portMapping = XRPortMapper(port: port)
		portMapping.mapTCP = true
		portMapping.mapUDP = false
		portMapping.desiredPublicPort = port
		self.portMapping = portMapping

		NotificationCenter.default.addObserver(
			self,
			selector: #selector(portMapperDidFinishWork(_:)),
			name: .XRPortMapperDidChanged,
			object: portMapping
		)

		if portMapping.open() == false {
			portMapperDidFinishWork(nil)
		}
	}

	@objc
	private func portMapperDidFinishWork(_: Notification?) {
		dispatchPrecondition(condition: .onQueue(.main))

		guard state == .listening, let portMapping else {
			return
		}

		NotificationCenter.default.removeObserver(
			self,
			name: .XRPortMapperDidChanged,
			object: portMapping
		)

		if portMapping.isMapped {
			directChatLogger.info("Direct chat: port \(self.hostPort, privacy: .public) mapped")
		} else {
			directChatLogger.error(
				"Direct chat: port mapping failed with error code \(portMapping.error, privacy: .public)"
			)
		}

		delegate?.directChatConnection(self, didStartListeningOnPort: hostPort)
	}

	@objc(sendMessage:)
	public func sendMessage(_ message: String) {
		sendLine(message)
	}

	@objc(sendAction:)
	public func sendAction(_ message: String) {
		sendLine(String(format: "%cACTION %@%c", 0x01, message, 0x01))
	}

	private func sendLine(_ line: String) {
		dispatchPrecondition(condition: .onQueue(.main))

		guard isConnected, let client else {
			return
		}

		/* A newline inside the text would be read by the peer as two
		 messages. The caller already split on newlines; this is a guard. */
		var sanitizedLine = line.replacingOccurrences(of: "\r", with: " ")
		sanitizedLine = sanitizedLine.replacingOccurrences(of: "\n", with: " ")

		guard let encoded = client.convert(toCommonEncoding: sanitizedLine) else {
			return
		}

		let data = NSMutableData(data: encoded)
		data.append("\r\n".data(using: .utf8)!)
		connection?.write(data as Data, timeout: writeTimeout)
	}

	private func consumeReceivedData(_ data: Data) {
		lineBuffer.append(data)

		while lineBuffer.length > 0 {
			let newlineRange = lineBuffer.range(
				of: Data([0x0A]),
				options: [],
				in: NSRange(location: 0, length: lineBuffer.length)
			)

			if newlineRange.location == NSNotFound {
				if lineBuffer.length > maximumLineLength {
					close(
						with: TDCFileTransferDialogSocket.error(
							withCode: .badParameter,
							description: "The peer sent a line which is too long"
						)
					)
				}
				return
			}

			let lineData = lineBuffer.subdata(with: NSRange(location: 0, length: newlineRange.location))
			lineBuffer.replaceBytes(
				in: NSRange(location: 0, length: newlineRange.location + 1),
				withBytes: nil,
				length: 0
			)

			consumeReceivedLine(lineData)

			if state == .closed {
				return
			}
		}
	}

	private func consumeReceivedLine(_ lineData: Data) {
		guard let client,
			var line = client.convert(fromCommonEncoding: lineData)
		else {
			return
		}

		line = line.trimmingCharacters(in: .newlines)

		if line.isEmpty {
			return
		}

		var isAction = false
		let actionPrefix = String(format: "%cACTION ", 0x01)

		if line.hasPrefix(actionPrefix) {
			let prefixLength = actionPrefix.count
			let suffixLength = line.hasSuffix(String(format: "%c", 0x01)) ? 1 : 0

			if line.count > (prefixLength + suffixLength) {
				let start = line.index(line.startIndex, offsetBy: prefixLength)
				let end = line.index(line.endIndex, offsetBy: -suffixLength)
				line = String(line[start..<end])
			} else {
				line = ""
			}

			isAction = true
		}

		delegate?.directChatConnection(self, didReceiveMessage: line, isAction: isAction)
	}

	// MARK: - Socket Delegate

	public func socket(_ socket: TDCFileTransferDialogSocket, didStartListeningOnPort port: UInt16) {
		dispatchPrecondition(condition: .onQueue(.main))

		guard socket === listeningServer else {
			return
		}

		hostPort = port
		mapListeningPort(port)
	}

	public func socket(_ socket: TDCFileTransferDialogSocket, didFailToListenWithError error: Error) {
		dispatchPrecondition(condition: .onQueue(.main))

		guard socket === listeningServer else {
			return
		}

		close(with: error)
	}

	public func socket(
		_ socket: TDCFileTransferDialogSocket,
		didAcceptConnection connection: TDCFileTransferDialogSocket
	) {
		dispatchPrecondition(condition: .onQueue(.main))

		if socket !== listeningServer || state != .listening {
			connection.disconnect()
			return
		}

		/* One peer only. Stop listening the moment somebody connects. */
		listeningServer = nil
		socket.disconnect()
		self.connection = connection
		connectionDidBecomeReady()
	}

	public func socketDidConnect(_ socket: TDCFileTransferDialogSocket) {
		dispatchPrecondition(condition: .onQueue(.main))

		guard socket === connection, state == .connecting else {
			return
		}

		connectionDidBecomeReady()
	}

	private func connectionDidBecomeReady() {
		state = .connected
		connection?.readData()
		delegate?.directChatConnectionDidConnect(self)
	}

	public func socket(_ socket: TDCFileTransferDialogSocket, didRead data: Data) {
		dispatchPrecondition(condition: .onQueue(.main))

		guard socket === connection, isConnected else {
			return
		}

		consumeReceivedData(data)

		if isConnected {
			socket.readData()
		}
	}

	public func socketDidWriteData(_: TDCFileTransferDialogSocket) {
		/* Nothing to do. Writes are fire and forget. */
	}

	public func socket(_ socket: TDCFileTransferDialogSocket, didDisconnectWithError error: Error?) {
		dispatchPrecondition(condition: .onQueue(.main))

		guard socket === connection || socket === listeningServer else {
			return
		}

		var closeError = error

		/* A clean close by the peer arrives as a "closed by peer" error.
		 Treat that as a normal end of the conversation. */
		if let nsError = error as NSError?,
			nsError.domain == TDCFileTransferDialogSocketErrorDomain,
			nsError.code == TDCFileTransferDialogSocketError.closedByPeer.rawValue
		{
			closeError = nil
		}

		close(with: closeError)
	}
}
